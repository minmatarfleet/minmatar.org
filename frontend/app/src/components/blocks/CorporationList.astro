---
import { i18n } from '@helpers/i18n'
const { t } = i18n(Astro.url)

import type { CorporationListData } from '@dtypes/layout_components';

import Flexblock from '@components/compositions/Flexblock.astro';
import FlexInline from '@components/compositions/FlexInline.astro';
import FixedFluid from '@components/compositions/FixedFluid.astro';
import TrashIcon from '@components/icons/buttons/TrashIcon.astro';
import FluidFixed from '@components/compositions/FluidFixed.astro';
import VerticalCenter from '@components/blocks/VerticalCenter.astro';
import Button from '@components/blocks/Button.astro';

interface Props {
    input_name?:        string;
    empty_text?:        string | boolean;
    readonly?:          boolean;
    corporations?:      CorporationListData[];
    [propName: string]: any;
}

const {
    input_name = 'corporations',
    empty_text = t('empty_corporations_text'),
    readonly = true,
    corporations = [],
    ...attributes
} = Astro.props;

delete attributes.class
---

<Flexblock
    class:list={['corporation-list', Astro.props.class]}
    {...attributes}
    gap='var(--space-s)'
    x-data={`{
        corporations: ${JSON.stringify(corporations)},
        corporations_ids: [],
        generate_ids() {
            this.corporations_ids = this.corporations.map(e => e.corporation_id).join(',')
        },
        ${readonly ? '' :
        `add_corporation(corporation) {
            if (!corporation) return

            const exist = this.corporations.find((existing_corporation) => existing_corporation.name === corporation.name)

            if (exist) return
            
            this.corporations.push({
                alt: corporation.name+' ${t('logo').toLowerCase()}',
                name: corporation.name,
                members_count: corporation.members_count,
                corporation_id: corporation.corporation_id,
            })

            this.generate_ids()
        },
        remove_corporation(name) {
            this.corporations = this.corporations.filter(item => item.name !== name)
            this.generate_ids()
        },`}
    }`}
    x-init="generate_ids"
>
    <FlexInline
        justification='space-between'
        gap="var(--space-s)"
    >
        <Flexblock gap="var(--space-3xs)">
            <h2>{t('corporations')}</h2>
            <small class="!mt-[var(--space-3xs)]" x-text={'`${corporations.length} '+t('corporation').toLowerCase()+'${corporations.length != 1 ? `s` : ``}`'}>0 {t('corporations').toLowerCase()}</small>
        </Flexblock>

        {!readonly && 
            <VerticalCenter>
                <Button
                    size='sm'
                    type="button"
                    class="[ w-full ]"
                    x-on:click="add_corporation(await show_corporations_finder())"
                >
                    {t('add_corporation')}
                </Button>
            </VerticalCenter>
        }
    </FlexInline>

    <Flexblock gap='var(--space-3xs)'>
        <template x-for="corporation in corporations">
            <FixedFluid width='64px' class="[ corporation-item ]">
                <picture>
                    <img loading="lazy" x-bind:src="get_corporation_logo(corporation.corporation_id, 256)" width="64" height="64" alt="corporation.alt" />
                </picture>
                <FluidFixed width='46px' breakpoint='60%'>
                    <VerticalCenter>
                        <Flexblock gap='0' class="[ w-full ]">
                            <span class="truncate text-[var(--highlight)]" x-text="corporation.name" />
                            <small class="truncate" x-text={`corporation.members_count+' ${t('members_count').toLowerCase()}'`} />
                        </Flexblock>
                    </VerticalCenter>
                    <VerticalCenter>
                        {!readonly &&
                            <Button
                                class="[ action-button ]"
                                x-on:click="remove_corporation(corporation.name)"
                                size='sm'
                                type="button"
                                narrow={true}
                            >
                                <TrashIcon slot="icon" />
                            </Button>
                        }
                    </VerticalCenter>
                </FluidFixed>
            </FixedFluid>
        </template>

        {empty_text &&
            <p x-bind:class="{ hidden: corporations.length > 0 }">{empty_text}</p>
        }
    </Flexblock>

    <input type="hidden" name={input_name} x-model="corporations_ids">
</Flexblock>

<style lang="scss">
    .corporation-list {
        transition: var(--fast-transition);

        img {
            object-fit: cover;
            width: 100%;
            height: 100%;
        }

        .corporation-item {
            @media (hover: hover) {
                .action-button {
                    transition: var(--fast-transition);
                    opacity: 0;
                }

                &:hover,
                &:focus-within {
                    .action-button {
                        opacity: 1;
                    }
                }
            }
        }

        &.disabled {
            pointer-events: none;
            opacity: 0.25;
        }
    }
</style>