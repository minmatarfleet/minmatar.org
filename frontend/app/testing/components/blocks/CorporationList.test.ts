import { experimental_AstroContainer as AstroContainer } from "astro/container";
import { expect, test, vi } from "vitest";
import CorporationList from "@components/blocks/CorporationList.astro";

test("CorporationList defaults", async () => {
  const container = await AstroContainer.create();
  const result = await container.renderToString(CorporationList, {});

  expect(result).toMatchSnapshot();
});
