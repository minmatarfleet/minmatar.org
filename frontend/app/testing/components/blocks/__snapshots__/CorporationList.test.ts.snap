// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CorporationList defaults 1`] = `
"<div class="flex-block corporation-list gap---space-s" x-data="{
        corporations: [],
        corporations_ids: [],
        generate_ids() {
            this.corporations_ids = this.corporations.map(e => e.corporation_id).join(',')
        },
        
    }" x-init="generate_ids" data-astro-cid-m6vonwzh="true" data-mod style="--flex_block_gap: var(--space-s);">  <div class="flex-inline" data-astro-cid-m6vonwzh="true" data-astro-cid-ud5fqf5k style="--flex_inline_gap: var(--space-s);--flex_inline_justification: space-between;">  <div class="flex-block gap---space-3xs" data-astro-cid-m6vonwzh="true" data-mod style="--flex_block_gap: var(--space-3xs);">  <h2 data-astro-cid-m6vonwzh>Corporations</h2> <small class="!mt-[var(--space-3xs)]" x-text="\`\${corporations.length} corporation\${corporations.length != 1 ? \`s\` : \`\`}\`" data-astro-cid-m6vonwzh>0 corporations</small>  </div>   </div>   <div class="flex-block gap---space-3xs" data-astro-cid-m6vonwzh="true" data-mod style="--flex_block_gap: var(--space-3xs);">  <template x-for="corporation in corporations" data-astro-cid-m6vonwzh> <div class="fixed-fluid [ corporation-item ]" data-astro-cid-m6vonwzh="true" data-mod data-astro-cid-34go4nvk style="--fixed_fluid_gap: var(--space-s);--fixed_fluid_width: 64px;--fixed_fluid_breakpoint: 50%;">  <picture data-astro-cid-m6vonwzh> <img loading="lazy" x-bind:src="get_corporation_logo(corporation.corporation_id, 256)" width="64" height="64" alt="corporation.alt" data-astro-cid-m6vonwzh> </picture> <div class="fluid-fixed" data-astro-cid-m6vonwzh="true" data-mod data-astro-cid-rwsocfon style="--fixed_fluid_gap: var(--space-s);--fixed_fluid_width: 46px;--fixed_fluid_breakpoint: 60%;">  <div class="!flex !items-center" data-astro-cid-m6vonwzh="true">  <div class="flex-block [ w-full ] gap-0" data-astro-cid-m6vonwzh="true" data-mod style="--flex_block_gap: 0;">  <span class="truncate text-[var(--highlight)]" x-text="corporation.name" data-astro-cid-m6vonwzh></span> <small class="truncate" x-text="corporation.members_count+' members count'" data-astro-cid-m6vonwzh></small>  </div>   </div> <div class="!flex !items-center" data-astro-cid-m6vonwzh="true">  </div>  </div>    </div>   </template> <p x-bind:class="{ hidden: corporations.length > 0 }" data-astro-cid-m6vonwzh>No corporation found.</p> </div>  <input type="hidden" name="corporations" x-model="corporations_ids" data-astro-cid-m6vonwzh>  </div>  "
`;
